src/transformers/models/phi3/modeling_phi3.py

```diff
@@ -0,0 +1,1595 @@
+        self.register_buffer("inv_freq", inv_freq, persistent=False)
+
+        # Build here to make `torch.jit.trace` work.
+        self._set_cos_sin_cache(
+            seq_len=max_position_embeddings, device=self.inv_freq.device, dtype=torch.get_default_dtype()
+        )
+
+    def _set_cos_sin_cache(self, seq_len, device, dtype):
+        self.max_seq_len_cached = seq_len
+        t = torch.arange(self.max_seq_len_cached, device=device, dtype=torch.int64).type_as(self.inv_freq)
+
+        freqs = torch.outer(t, self.inv_freq)
+        # Different from paper, but it uses a different permutation in order to obtain the same calculation
+        emb = torch.cat((freqs, freqs), dim=-1)
+        self.register_buffer("cos_cached", emb.cos().to(dtype), persistent=False)
+        self.register_buffer("sin_cached", emb.sin().to(dtype), persistent=False)
+
+    def forward(self, x, seq_len=None):
+        # x: [bs, num_attention_heads, seq_len, head_size]
+        if seq_len > self.max_seq_len_cached:
+            self._set_cos_sin_cache(seq_len=seq_len, device=x.device, dtype=x.dtype)
+
+        return (
+            self.cos_cached[:seq_len].to(dtype=x.dtype),
+            self.sin_cached[:seq_len].to(dtype=x.dtype),
+        )
+
+
+class _Phi3ScaledRotaryEmbedding(nn.Module):
+    def __init__(
+        self,
+        dim,
+        short_factor,
+        long_factor,
+        max_position_embeddings=2048,
+        original_max_position_embeddings=2048,
+        base=10000,
+    ):
+        super().__init__()
+
+        self.dim = dim
+        self.short_factor = short_factor
+        self.long_factor = long_factor
+        self.max_position_embeddings = max_position_embeddings
+        self.original_max_position_embeddings = original_max_position_embeddings
+        self.base = base
+
+    def _calc_mscale(self, scale):
+        raise NotImplementedError("`_calc_mscale` should be implemented in subclasses")
+
+    @torch.no_grad()
+    def forward(self, x, seq_len=None):
+        if seq_len is None:
+            seq_len = x.shape[-2]
+        t = torch.arange(seq_len, device=x.device, dtype=torch.float32)
+
+        if seq_len > self.original_max_position_embeddings:
+            t = torch.arange(seq_len, device=x.device, dtype=torch.float32)
+            rescale_factors = torch.tensor(self.long_factor, dtype=torch.float32, device=x.device)
+        else:
+            t = torch.arange(self.original_max_position_embeddings, device=x.device, dtype=torch.float32)
+            rescale_factors = torch.tensor(self.short_factor, dtype=torch.float32, device=x.device)
+
+        inv_freq = 1.0 / (
+            rescale_factors * (self.base ** (torch.arange(0, self.dim, 2).float().to(x.device) / self.dim))
+        )
+
+        freqs = torch.outer(t, inv_freq)
+        mscale = self._calc_mscale(self.max_position_embeddings / self.original_max_position_embeddings)
+        emb = torch.cat((freqs, freqs), dim=-1)
```

Comment:
this should be equivalent to scaling the positional ids no? (as long as the inv freq is recomputed)